### ISSUE TITLE
Unresolved symbol while pickling quote

### ISSUE THREAD

**Issue by nicolasstucki** [lampepfl/dotty#13732]
## Compiler version

3.1.0

## Minimized code

```Scala
import scala.quoted.*

def generateImpl(using q: Quotes): Expr[Unit] =
  '{ def runEffect[T]: T = ${ runEffectImpl[T] } }

inline def runEffectImpl[T: Type]: Expr[T] = ???
```

Related to #13563

## Output

```scala
java.lang.AssertionError: assertion failed: unresolved symbols: type T (line 3) #13947 when pickling tests/pos-macros/i13563b.scala
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:777)
        at dotty.tools.dotc.quoted.PickledQuotes$.pickle(PickledQuotes.scala:167)
        at dotty.tools.dotc.quoted.PickledQuotes$.pickleQuote(PickledQuotes.scala:36)
        ...
```

## Expectation

Should compile

**Comment by nicolasstucki** [comment]
Workaround 
```diff
- inline def runEffectImpl[T: Type]: Expr[T] = ???
+ def runEffectImpl[T: Type]: Expr[T] = ???
```

**Comment by griggt** [comment]
@deniszo and I looked at this during today's spree.  Some observations:

 - The [workaround above](https://github.com/lampepfl/dotty/issues/13732#issuecomment-940954160) does not seem to help on 3.1.0 or master
 - The example compiles with #12540, but fails -Ycheck
 - The workaround above *does* seem to work on the #12540 branch, with -Ycheck
 - The example compiles if the `Type` context bound is on both `runEffect` and `runEffectImpl`, or on neither
 - The example compiles if `runEffectImpl` is called with a type parameter of `Nothing` (inferred or explicit)

It seems that for some reason in this case, `transformQuotation` 

https://github.com/lampepfl/dotty/blob/b311b4ecf84bf33dd387353a7dfb989e73f236ca/compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala#L129-L141

wants to pickle a quote whose body consists solely of `TypeTree[TypeRef(NoPrefix,type T)]`, and I get the impression that is somewhat nonsensical to pickle in isolation?

**Comment by griggt** [comment]
Is it expected that this issue can be solved separately from / without #12540 (or a similar reworking of quote pickling)?
